hp-laptop-kali-linux build chain: hf-hub feature shape blocks rustls migration

## Observed during 2026-05-18 multi-host mesh deploy

The Kali Linux host (Tailscale `100.82.83.122`) has no passwordless sudo, so installing `libssl-dev` to satisfy the default `openssl-sys` build is not an option from a remote-only deploy script. Attempted workaround: override `tirami-infer/Cargo.toml` to pull `hf-hub` with `default-features = false, features = ["tokio", "rustls-tls"]`. That removed `tokio-native-tls` + `ureq` from the dep graph (good) but broke `tirami-infer` source:

```
error[E0432]: unresolved import `hf_hub::api::sync`
   --> crates/tirami-infer/src/...
```

`hf-hub`'s `sync` API is exposed under a feature that's gated to `ureq`, which in turn pulls `native-tls` → `openssl-sys`. Migration is therefore source-level, not just Cargo.toml:

## Two viable fixes (separate PR scope)

1. **Replace `hf_hub::api::sync` callsites with the `tokio` async API** inside `tirami-infer`. This is the larger source change but cleanest dep graph (drops `ureq` entirely).
2. **Pin `hf-hub` 0.4.x with `features = ["tokio", "ureq", "rustls-tls"]`** AND ensure `ureq` 2.x has its `tls` feature actually pointing to rustls. May still require a `[patch.crates-io]` for `ureq`.

## Impact today

- macOS hosts (M-series + Intel) build fine with the default Cargo.toml (libssl bundled by `brew`-installed openssl or autodetected via `OPENSSL_DIR`).
- Linux hosts WITH passwordless sudo to install `libssl-dev` build fine (verified on `asusrogflowx13mint`, Linux 6.17 x86_64).
- Linux hosts WITHOUT sudo cannot join the mesh until one of the above fixes lands.

## Suggested ordering

Bundle this with #137 (transitive CVE in `iroh-relay` deps): both want the protocol to be rustls-native end-to-end, which would also retire the `hickory-proto`-derived NSEC3 CVE path.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -101,7 +101,11 @@ llama-cpp-2 = "0.1"
 
 # Tokenization
 tokenizers = { version = "0.21", default-features = false, features = ["onig"] }
-hf-hub = "0.4"
+# Phase 25 #138 — drop the default `ureq + native-tls` path so
+# Linux hosts without `libssl-dev` (e.g. immutable container, no
+# passwordless sudo) can build. We use the tokio async surface +
+# rustls TLS and block_on at the (sync) caller boundary.
+hf-hub = { version = "0.4", default-features = false, features = ["tokio", "rustls-tls"] }
 
 # Networking (Iroh)
 iroh = { version = "0.97", features = ["address-lookup-mdns"] }

--- a/crates/tirami-infer/src/model_registry.rs
+++ b/crates/tirami-infer/src/model_registry.rs
@@ -84,8 +84,31 @@ pub struct ResolvedModel {
 }
 
 /// Download (or find in cache) model files from HuggingFace.
+///
+/// Phase 25 #138 — `hf-hub`'s sync API (`hf_hub::api::sync`) is
+/// gated to the `ureq + native-tls` features, which require
+/// libssl-dev on Linux. We use the tokio async surface with
+/// rustls-tls and bridge sync→async via an isolated runtime on
+/// a worker thread. The worker-thread isolation avoids "cannot
+/// start a runtime from within a runtime" panics when the caller
+/// is already inside a tokio context (e.g. `tirami node` boot).
 pub fn resolve_model(spec: &ModelSpec) -> Result<ResolvedModel, TiramiError> {
-    use hf_hub::api::sync::ApiBuilder;
+    let spec = spec.clone();
+    std::thread::scope(|s| {
+        s.spawn(move || {
+            let rt = tokio::runtime::Builder::new_current_thread()
+                .enable_all()
+                .build()
+                .map_err(|e| TiramiError::ModelLoadError(format!("tokio rt: {e}")))?;
+            rt.block_on(resolve_model_async(&spec))
+        })
+        .join()
+        .unwrap_or_else(|_| Err(TiramiError::ModelLoadError("worker thread panicked".into())))
+    })
+}
+
+async fn resolve_model_async(spec: &ModelSpec) -> Result<ResolvedModel, TiramiError> {
+    use hf_hub::api::tokio::ApiBuilder;
 
     let api = ApiBuilder::new()
         .with_progress(true)
@@ -97,11 +120,13 @@ pub fn resolve_model(spec: &ModelSpec) -> Result<ResolvedModel, TiramiError> {
     let model_path = api
         .model(spec.gguf_repo.clone())
         .get(&spec.gguf_file)
+        .await
         .map_err(|e| TiramiError::ModelLoadError(format!("download GGUF: {e}")))?;
 
     let tokenizer_path = api
         .model(spec.tokenizer_repo.clone())
         .get(&spec.tokenizer_file)
+        .await
         .map_err(|e| TiramiError::ModelLoadError(format!("download tokenizer: {e}")))?;
 
     Ok(ResolvedModel {
@@ -291,7 +316,26 @@ fn download_hf_file(
     file: &str,
     _branch: &str,
 ) -> Result<(PathBuf, PathBuf), TiramiError> {
-    use hf_hub::api::sync::ApiBuilder;
+    let repo = repo.to_string();
+    let file = file.to_string();
+    std::thread::scope(|s| {
+        s.spawn(move || {
+            let rt = tokio::runtime::Builder::new_current_thread()
+                .enable_all()
+                .build()
+                .map_err(|e| TiramiError::ModelLoadError(format!("tokio rt: {e}")))?;
+            rt.block_on(download_hf_file_async(&repo, &file))
+        })
+        .join()
+        .unwrap_or_else(|_| Err(TiramiError::ModelLoadError("worker thread panicked".into())))
+    })
+}
+
+async fn download_hf_file_async(
+    repo: &str,
+    file: &str,
+) -> Result<(PathBuf, PathBuf), TiramiError> {
+    use hf_hub::api::tokio::ApiBuilder;
 
     let api = ApiBuilder::new()
         .with_progress(true)
@@ -301,10 +345,11 @@ fn download_hf_file(
     let model_repo = api.model(repo.to_string());
     let model_path = model_repo
         .get(file)
+        .await
         .map_err(|e| TiramiError::ModelLoadError(format!("download {repo}/{file}: {e}")))?;
 
     // Try tokenizer.json from the GGUF repo first.
-    let tokenizer_path = match model_repo.get("tokenizer.json") {
+    let tokenizer_path = match model_repo.get("tokenizer.json").await {
         Ok(p) => p,
         Err(_) => {
             // Fallback: derive the non-GGUF repo name (strip -GGUF suffix).
@@ -315,7 +360,7 @@ fn download_hf_file(
                 .to_string();
             if non_gguf_repo != repo {
                 let alt_repo = api.model(non_gguf_repo);
-                alt_repo.get("tokenizer.json").map_err(|e| {
+                alt_repo.get("tokenizer.json").await.map_err(|e| {
                     TiramiError::ModelLoadError(format!("download tokenizer: {e}"))
                 })?
             } else {

